src/Microsoft.DotNet.Darc/src/DarcLib/Actions/Remote.cs

```diff
@@ -281,41 +281,320 @@ public async Task MergePullRequestAsync(string pullRequestUrl, MergePullRequestP
+            {
+                // Nothing to do.
+                return toUpdate;
+            }
+
+            DependencyGraphBuildOptions dependencyGraphBuildOptions = new DependencyGraphBuildOptions()
+            {
+                IncludeToolset = true,
+                LookupBuilds = true,
+                NodeDiff = NodeDiff.None
+            };
+
+            // Now make a walk over coherent dependencies. Note that coherent dependencies could make
+            // a chain (A->B->C). In all cases we need to walk to the head of the chain, keeping track
+            // of all elements in the chain. Also note that we are walking all dependencies here, not
+            // just those that match the incoming AssetData and aligning all of these based on the coherency data.
+            Dictionary<string, DependencyGraphNode> nodeCache = new Dictionary<string, DependencyGraphNode>();
+            HashSet<DependencyDetail> visited = new HashSet<DependencyDetail>();
+            foreach (DependencyDetail dependency in leavesOfCoherencyTrees)
+            {
+                // If the dependency was already updated, then skip it (could have been part of a larger
+                // dependency chain)
+                if (visited.Contains(dependency))
+                {
+                    continue;
+                }
+
+                // Walk to head of dependency tree, keeping track of elements along the way.
+                // If we hit a pinned dependency in the walk, that means we can't move
+                // the dependency and therefore it is effectively the "head" of the subtree.
+                // We will still visit all the elements in the chain eventually in this algorithm:
+                // Consider A->B(pinned)->C(pinned)->D.
+                List<DependencyDetail> updateList = new List<DependencyDetail>();
+                DependencyDetail currentDependency = dependency;
+                while (!string.IsNullOrEmpty(currentDependency.CoherentParentDependencyName) && !currentDependency.Pinned)
+                {
+                    updateList.Add(dependency);
+                    DependencyDetail parentCoherentDependency = dependencies.FirstOrDefault(d =>
+                        d.Name.Equals(currentDependency.CoherentParentDependencyName, StringComparison.OrdinalIgnoreCase));
+                    // If the coherent dependency is not found, this is non-valid.
+                    if (parentCoherentDependency == null)
+                    {
+                        throw new DarcException($"Dependency {currentDependency.Name} has non-existent parent " +
+                            $"dependency {currentDependency.CoherentParentDependencyName}");
+                    }
+                    currentDependency = parentCoherentDependency;
+                }
+
+                if (!updateList.Any())
+                {
+                    throw new DarcException($"Unexpected error while determining coherent dependencies to update.");
```

Comment:
>Unexpected error while determining coherent dependencies to update. [](start = 46, length = 67)

Why is "the list is empty" an "unexpected error"... what does it mean if this list is empty?

Reply:
It was more an internal consistency check that the calculation of coherency tree roots was valid.  I Think it can just be removed.